keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,884 @@
+single authority across all clusters. It is that authority’s responsibility to
+ensure that a name is shared by multiple services within the namespace if and
+only if they are instances of the same service.
+
+Most information about the service, including ports, backends and topology, will
+continue to be stored in the Service object, which is name mapped to the service
+export.
+
+#### Restricting Exports ####
+
+Cluster administrators may use RBAC rules to prevent creation of
+`ServiceExports` in select namespaces. While there are no general restrictions
+on which namespaces are allowed, administrators should be especially careful
+about permitting exports from `kube-system` and `default`. As a best practice,
+admins may want to tightly or completely prevent exports from these namespaces
+unless there is a clear use case.
+
+### Exported Service Behavior Expectations
+
+#### SuperclusterIP
+
+When a `ServiceExport` is created, an IP address is reserved and assigned to
+this supercluster `Service`. This IP may be supercluster-wide, or assigned on a
+per-cluster basis. Requests to the corresponding IP from within a given cluster
+will route to endpoint addresses for the aggregated Service. 
+
+Note: this doc does not discuss `NetworkPolicy`, which cannot currently be used
+to describe a policy that applies to a multi-cluster service.
+
+#### DNS
+
+When a `ServiceExport` is created, this will cause a domain name for the
+multi-cluster service to become accessible from within the supercluster. The
+domain name will be
+`<service-export-name>.<service-export-namespace>.svc.supercluster.local`.
+Requests to this domain name from within the supercluster will resolve to the
+supercluster VIP, which points to the endpoint addresses for pods within the
+underlying `Service`(s) across the supercluster. All service consumers must use
+the `*.svc.supercluster.local` name to enable supercluster routing, even if
+there is a matching `Service` with the same namespaced name in the local
+cluster. There will be no change to existing behavior of the `svc.cluster.local`
+zone.
+
+#### EndpointSlice
+
+When a `ServiceExport` is created, this will cause `EndpointSlice` objects for
+the underlying `Service` to be created in each cluster within the supercluster.
+One or more `EndpointSlice` resources will exist for each cluster that exported
+the `Service`, with each `EndpointSlice` containing only endpoints from its
+source cluster. These `EndpointSlice` objects will be marked as managed by the
+supercluster service controller, so that the endpoint slice controller doesn’t
+delete them.
+
+```
+<<[UNRESOLVED]>>
+We have not yet sorted out scalability impact here. We hope the upper bound for
+imported endpoints + in-cluster endpoints will be ~= the upper bound for
+in-cluster endpoints today, but this remains to be determined.
+<<[/UNRESOLVED]>>
+```
+
+#### Endpoint TTL
+
+To prevent stale endpoints from persisting in the event that a cluster becomes
```

Comment:
another important problem that use to happen in these scenarios (commonly with BGP peering) is unstability or flapping due to problems with the peers, per example because of network hiccups or other connectivity problems. If it connects and reconnects constantly it creates a lot of pressure on the clusters creating and deleting resources. Maybe we should indicate that we should add some protection to avoid this?

I don't think that right now implementations details are necessary but just for reference and completeness,   BGP solves it using flap damping https://tools.ietf.org/html/rfc2439, but just whatever mechanism that guarantees certain stability will be valid

Reply:
We've purposefully avoided specifying exactly how the supercluster reconciles services between clusters beyond what's necessary to implement the API, but here's an example of how I see this working:

The controller periodically validates connectivity to each cluster in the supercluster through some mechanism, then in each cluster it manages (in a centralized model it would be all clusters, in a distributed model maybe just the local cluster) it would renew the source-cluster-specific lease for each currently connected cluster. The TTL would be some multiple of the validation interval and allow for some flakiness. This way we'd only cleanup endpoints if a source cluster was unreachable for a sustained period.

For the reasons you suggest, I don't think a supercluster controller would want to explicitly delete endpoints for a cluster because it was unreachable - rather let the lease expire in that case. This way there should be no action if there's a temporary disconnect so long as it recovered before the TTL.

What if this instead read:

To prevent stale endpoints from persisting in the event that a cluster becomes
unreachable to the supercluster controller for a sustained period of time, ...

Does that make more sense?

Reply:
yeah, I was more concerned about a possible overlook of this problem, but is obvious you have it very clear and well designed 😅 . Current wording works for me too. Thanks for the clarification

Reply:
My big concern here is to quickly weed out endpoints that are not alive.  absent real health-checking, this is hard.

Reply:
We should figure out how to do better than TTL (see DNS caching)

Reply:
I've been thinking about this, and maybe TTL doesn't need to be part of this spec and instead made a responsibility of the implementation. Even as described, TTL only handles the worst case scenario should a cluster or supercluster controller become unavailable or otherwise stop functioning. That's clearly necessary, but there are various ways to go about it.

An implementation should strive to do better than this - for example, quickly cleaning up endpoints that have been removed from their source cluster and/or actually perform health checking. I don't think it's necessary (or desirable) to define that behavior as preference may depend on the environment or other to prioritize other goals.

Reply:
Absent a kubernetes-level healthcheck spec, implementations are free to do whatever they want, but there's no consistency.  Adding an HC block to Services may be the "real" solution, but that's probably out-of-scope for this KEP.

Reply:
Agreed. That said, what are thoughts on keep/removing TTL as part of the spec?

- We keep it to bound worst-case endpoint cleanup behavior for all implementations, but with the understanding that implementations should strive to do better. This still means inconsistent behavior but at least guarantees that endpoints wont live forever.
- We remove it and leave cleanup entirely to implementation. There's even less consistency and we open to the possibility of technically compliant implementations that leave dead endpoints forever. 

It seems like a useful thing to have right now, but if we expect e.g. an HC block to be added to Services TTL may become redundant quickly as MCS adopts that instead